### DISCUSSION TITLE
jsonable_encoder should handle "nan"

### DISCUSSION THREAD

**Discussion by henninglebbaeus** [fastapi/fastapi#8029] [Questions] [answered]
## Description

I have a route that simply fetches objects from a database and returns a list of them. They are instances of a pydantic model. However, when such an instance contains a property whose value is `nan`, fastapi throws an error.
```
ValueError: Out of range float values are not JSON compliant
```
I was able to track it down to the `starlette.response.JSONResponse` class, that under the hood calls `json.dumps` with the argument `allow_nan=False`.

The result is that if a value is `nan`, the json dump fails here. However, this passes the `jsonable_encoder` silently, as it returns objects of type `float` as is. 

May I suggest to implement a check whether a float is `nan` and transform it to something else, maybe `None`? I am happy to send a PR if you think it's worthwhile.

### Example Code
```Python
class Model(pydantic.BaseModel):
    value: float

app.get('/this')
def get_this():
    x = Model(value=float('nan'))
    return x
```

### System Info
Linux 42-Ubuntu SMP Wed Aug 7 15:17:54 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
Python 3.7.3
fastapi                  0.31.0
starlette                0.12.0

**Comment by zbarnett** [comment, marked as answer]
I ran into the exact same issue and solved it by replacing the built-in JSON library with a custom one... ([orjson](https://github.com/ijl/orjson) worked well for my needs)

```python
class ORJSONResponse(JSONResponse):
    media_type = "application/json"

    def render(self, content: typing.Any) -> bytes:
        return orjson.dumps(content)
```

...and then overriding the default response class.
```python
app = FastAPI(default_response_class=ORJSONResponse)
```

Keep in mind the default_response_class option was added in [FastAPI 0.39.0](https://github.com/tiangolo/fastapi/releases/tag/0.39.0)

**Comment by dmontagu** [comment]
In addition to @zbarnett 's great suggestion, in pydantic v1 (which is coming very soon) there will be better support for specifying a custom encoder for your pydantic models. So that may (eventually) help from another angle.

**Comment by henninglebbaeus** [comment]
Thanks, @zbarnett . Your solution is actually much nicer then what I had in mind.

In the end, we moved the NaN checks to the data ingestion part outside of fastapi.

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue @henninglebbaeus  :+1:

**Comment by lanetao** [comment]
> I ran into the exact same issue and solved it by replacing the built-in JSON library with a custom one... ([orjson](https://github.com/ijl/orjson) worked well for my needs)
> 
> ```python
> class ORJSONResponse(JSONResponse):
>     media_type = "application/json"
> 
>     def render(self, content: typing.Any) -> bytes:
>         return orjson.dumps(content)
> ```
> 
> ...and then overriding the default response class.
> 
> ```python
> app = FastAPI(default_response_class=ORJSONResponse)
> ```
> 
> Keep in mind the default_response_class option was added in [FastAPI 0.39.0](https://github.com/tiangolo/fastapi/releases/tag/0.39.0)

good solution and ths.

**Comment by Dhavaleswar** [comment]
What if request has null values, 

```
class ReturnSeries(BaseModel):
       time_series: List[float]
```

I am getting below error, if input has nan values

`requests.exceptions.InvalidJSONError: Out of range float values are not JSON compliant
`

**Comment by zljubisic** [comment]
What is wrong with converting NaN to null?
I manually replaced NaN's with null and such json is perfectly fine.

```
dct = json.loads("""[
  {
    "idem": "A",
    "price": null
  },
  {
    "idem": "B",
    "price": 1
  }
]""")
dct
[{'idem': 'A', 'price': None}, {'idem': 'B', 'price': 1}]
```

**Comment by michealroberts** [comment]
I think issues like this is why I am more inclined to stick with Django for all of its flaws, the manual process and steps I have gone through to setup an API only to see errors like this are purely frustrating, whereas with Django, it's soooo much DRY-er.

**Comment by lwiklendt** [comment]
> I ran into the exact same issue and solved it by replacing the built-in JSON library with a custom one... ([orjson](https://github.com/ijl/orjson) worked well for my needs)
> 
> ```python
> class ORJSONResponse(JSONResponse):
>     media_type = "application/json"
> 
>     def render(self, content: typing.Any) -> bytes:
>         return orjson.dumps(content)
> ```
> 
> ...and then overriding the default response class.
> 
> ```python
> app = FastAPI(default_response_class=ORJSONResponse)
> ```
> 
> Keep in mind the default_response_class option was added in [FastAPI 0.39.0](https://github.com/tiangolo/fastapi/releases/tag/0.39.0)

`ORJSONResponse` is included in FastAPI so you can just use it directly:

```python
app = FastAPI(default_response_class=fastapi.responses.ORJSONResponse)
```

**Reply by uxduck** [reply]
You still need to install orjson. FastAPI doesn't install it:
```
    assert orjson is not None, "orjson must be installed to use ORJSONResponse"
           ^^^^^^^^^^^^^^^^^^
AssertionError: orjson must be installed to use ORJSONResponse
```

**Reply by kdoda-cipher** [reply]
I ran into the same issue. You need to install the package.

**Comment by JavierSanchezCastro** [comment]
What about passing NaN as input body parameter?

Opened discussion here: https://github.com/tiangolo/fastapi/discussions/10141
